---
title: "CUT&TAG Processing Pipeline"
author:
- name: Mireia Ramos-RodrÃ­guez
  affiliation: Department of Experimental and Health Sciences, University Pompeu Fabra
  email: mireia.ramos@upf.edu
package: pipelineNGS
output:
  BiocStyle::html_document
abstract: |
  This vignette describes the pipeline implemented to analyze CUT&TAG data.
vignette: |
  %\VignetteIndexEntry{CUT&TAG Processing Pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

[C&T description]

```{r ct-description, echo=FALSE, fig.cap="Differences between immunoprecipitation and in antibody-targeted chromatin profiling strategies. A. ChIP-seq experimental procedure. B. CUT&Tag experimental procedure. Cells and nuclei are indicated in grey, chromatin as red nucleosomes, and a specific chromatin protein in green."}
knitr::include_graphics("https://yezhengstat.github.io/CUTTag_tutorial/figures/ChIPseqCUTTag.png")
```

Recommended CUT&TAG pipeline: https://yezhengstat.github.io/CUTTag_tutorial/

# Pre-requisites

- **Bowtie2**[^bowtie2] (version 2.4.1).
- **Samtools**[^samtools] (version 1.10).
- **Bedtools**[^bedtools] (version 2.29.2).
- **SEACR**[^SEACR] (version 1.3).

[^bowtie2]: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
[^samtools]: http://www.htslib.org/
[^bedtools]: https://bedtools.readthedocs.io/en/latest/
[^SEACR]: https://github.com/FredHutch/SEACR

# Pipeline description

## FastQ Quality Control with FastQC


## Aligment with Bowtie2

```
bowtie2 -t -x ~/refs/indexes/hg38 -1 /path/to/fastq/sample_1_read1.fastq.gz -2 /path/to/fastq/sample_1_read2.fastq.gz -p 5 2> Logs/sample_1.alignment.log | samtools view - -b -@ 4 | samtools sort - -@ 4 -m 2G -o BAM/sample_1.raw.bam ; samtools index BAM/sample_1.raw.bam -@ 4 ; samtools idxstats BAM/sample_1.raw.bam > Logs//sample_1.raw.idxstats.log
```

## Filtering out unnecessary reads

Remove unmapped and blacklisted reads:

```
samtools idxstats BAM/10759_1.raw.bam | cut -f 1 | grep -v chrM | grep -v chrUn | grep -v _random | grep -v _hap | grep -v _gl | grep -v EBV | xargs samtools view 
-b BAM/10759_1.raw.bam -@ 4 -F 4 | samtools view - -b -L ~/refs/Blacklist/lists/hg38-blacklist.v2.bed -U BAM/10759_1.tmp.bam -o trash.bam 
```

Fixmates

```
samtools sort -n BAM/10759_1.tmp.bam -@ 4 -m 2G -o - | samtools fixmate -m - - | samtools sort - -@ 4 -m 2G -o BAM/10759_1.fixmate.bam 
```

Remove duplicates

```
samtools markdup BAM/10759_1.fixmate.bam BAM/10759_1.bam -r -s 2> Logs/10759_1.rmdup.log ; samtools index BAM/10759_1.bam -@ 4 ; samtools idxstats BAM/10759_1.bam 
```

## Peak calling with SEACR

Convert to appropriate format

```
# Convert to BED
samtools sort -n BAM/sample_1.bam -@ 4-m 2G -o - | bedtools bamtobed -bedpe -i stdin | awk '$1==$4 && $6-$2 < 1000 {print $0}' | cut -f 1,2,6 | sort -dsk1,1 -k2n,2 -k3nr,3 > BAM/sample_1.bed
# Convert to BedGraph
bedtools genomecov -bg -i sample_1.bed -g ~/refs/hg38.chrom.sizes > sample_1.bedgraph
```

Peak calling

```
samtools sort -n BAM/sample_1.bam -@ 4-m 2G -o - | bedtools bamtobed -bedpe -i stdin | awk '$1==$4 && $6-$2 < 1000 {print $0}' | cut -f 1,2,6 | sort -dsk1,1 -k2n,2 -k3nr,3 > BAM/sample_1.bed
bedtools genomecov -bg -i sample_1.bed -g ~/refs/hg38.chrom.sizes > sample_1.bedgraph
```

# Running the pipeline

## Sequentially


## Step-by-step



# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
