---
title: "CUT&TAG Processing Pipeline"
author:
- name: Mireia Ramos-RodrÃ­guez
  affiliation: Department of Experimental and Health Sciences, University Pompeu Fabra
  email: mireia.ramos@upf.edu
package: pipelineNGS
output:
  BiocStyle::html_document
abstract: |
  This vignette describes the pipeline implemented to analyze CUT&TAG data.
vignette: |
  %\VignetteIndexEntry{CUT&TAG Processing Pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval = FALSE,
                      fig.algn = "center",
                      warning = FALSE,
                      message = FALSE,
                      fig.width=4, 
                      fig.height=4)

```


# Introduction

The mapping of chromatin features genome-wide has traditionally been performed using chromatin immunoprecipitation (ChIP), in which chromatin is cross-linked and solubilized, and an antibody to a protein or modification of interest is used to immunoprecipitate the bound DNA (Fig. 1a).

An alternative chromatin profiling strategy is enzyme tethering in situ whereby the chromatin protein or modification of interest is targeted by an antibody or fusion protein. Then, the underlying DNA is marked or cleaved, and a succession of enzyme-tethering methods have been introduced over the past two decades. Cleavage Under Targets & Tagmentation (CUT&Tag) is a tethering method that uses a protein-A-Tn5 (pA-Tn5) transposome fusion protein (Fig. 1b).

```{r ct-description, echo=FALSE, eval=TRUE, fig.cap="Differences between immunoprecipitation and in antibody-targeted chromatin profiling strategies. A. ChIP-seq experimental procedure. B. CUT&Tag experimental procedure. Cells and nuclei are indicated in grey, chromatin as red nucleosomes, and a specific chromatin protein in green."}
knitr::include_graphics("https://yezhengstat.github.io/CUTTag_tutorial/figures/ChIPseqCUTTag.png")
```

This package implements some functions to facilitate processing of CYT&TAG data, based on the original published CUT&TAG analysis tutorial.

> Original **CUT&TAG pipeline**: https://yezhengstat.github.io/CUTTag_tutorial/

# Pre-requisites

- **Bowtie2**[^bowtie2] (version 2.4.1).
- **Samtools**[^samtools] (version 1.10).
- **Bedtools**[^bedtools] (version 2.29.2).
- **SEACR**[^SEACR] (version 1.3).

[^bowtie2]: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
[^samtools]: http://www.htslib.org/
[^bedtools]: https://bedtools.readthedocs.io/en/latest/
[^SEACR]: https://github.com/FredHutch/SEACR

# Pipeline description

## FastQ Quality Control with FastQC

```
fastqc -t 5 -o FastQC/ /path/to/fastq/sample_1_read1.fastq.gz
```

The **FAIL due to discordant sequence content** at the beginning of the reads are a **common phenomenon for CUT&Tag reads** and is likely related to a Tn5 sequence preference. Authors do not recommend read trimming, as the bowtie2 parameters that they list will give accurate mapping information without trimming.



## Aligment with Bowtie2

```
bowtie2 -t -x ~/refs/indexes/hg38 -1 /path/to/fastq/sample_1_read1.fastq.gz -2 /path/to/fastq/sample_1_read2.fastq.gz -p 5 2> Logs/sample_1.alignment.log | samtools view - -b -@ 4 | samtools sort - -@ 4 -m 2G -o BAM/sample_1.raw.bam ; samtools index BAM/sample_1.raw.bam -@ 4 ; samtools idxstats BAM/sample_1.raw.bam > Logs//sample_1.raw.idxstats.log
```

This command is optimized for obtaining very sensitive alignments without the need of trimming adapters from the input reads.

## Filtering out unnecessary reads

Remove non-cannonical, unmapped and blacklisted reads:

```
samtools idxstats BAM/10759_1.raw.bam | cut -f 1 | grep -v chrM | grep -v chrUn | grep -v _random | grep -v _hap | grep -v _gl | grep -v EBV | xargs samtools view 
-b BAM/10759_1.raw.bam -@ 4 -F 4 | samtools view - -b -L ~/refs/Blacklist/lists/hg38-blacklist.v2.bed -U BAM/10759_1.tmp.bam -o trash.bam 
```

Fixmates:

```
samtools sort -n BAM/10759_1.tmp.bam -@ 4 -m 2G -o - | samtools fixmate -m - - | samtools sort - -@ 4 -m 2G -o BAM/10759_1.fixmate.bam 
```

Remove duplicates:

```
samtools markdup BAM/10759_1.fixmate.bam BAM/10759_1.bam -r -s 2> Logs/10759_1.rmdup.log ; samtools index BAM/10759_1.bam -@ 4 ; samtools idxstats BAM/10759_1.bam 
```

## Peak calling with SEACR

Convert to appropriate format:

```
# Convert to BED
samtools sort -n BAM/sample_1.bam -@ 4-m 2G -o - | bedtools bamtobed -bedpe -i stdin | awk '$1==$4 && $6-$2 < 1000 {print $0}' | cut -f 1,2,6 | sort -dsk1,1 -k2n,2 -k3nr,3 > BAM/sample_1.bed
# Convert to BedGraph
bedtools genomecov -bg -i sample_1.bed -g ~/refs/hg38.chrom.sizes > sample_1.bedgraph
```

Peak calling:

```
SEACR_1.3.sh BAM/sample_1.bam 0.01 norm stringent Peaks/sample_1.peaks.0.01
```

# Running the pipeline

- For paired-end samples, the input `fastq_files` should be a list of length equal to the number of samples, with each element containing 2 file paths (read 1 and read 2).

```{r}
samples <- read.delim(here::here("data/sample_info.tsv"))
fastq <- split(samples$files, samples$sampleID) # Create a list with each sample containing the two read files (R1 and R2)
```


## Each step for all samples

By default, the function `processCUTTAG` runs each step of the pipeline for each sample, before jumping to the next step. 

```{r}
processCUTTAG(fastq_files=fastq,
              out_name=names(fastq),
              MoreArgs=list(index="~/refs/indexes/hg38",
                            blacklist="~/refs/Blacklist/lists/hg38-blacklist.v2.bed",
                            chr_sizes="~/refs/hg38.chromSizes.txt",
                            cores=5,
                            seacr_type = "stringent",
                            seacr_top = 0.01))
```


## One full sample at a time

In case you want to run all the pipeline one sample at a time, you can take advantage of the `mapply` function, which will run `processCUTTAG` for each element of the provided list.`

```{r}
mapply(processCUTTAG,
       fastq_files=fastq,
       out_name=names(fastq),
       MoreArgs=list(index="~/refs/indexes/hg38",
                     blacklist="", # Do not remove blacklisted reads
                     chr_sizes="~/refs/hg38.chromSizes.txt",
                     cores=5,
                     seacr_type = "stringent",
                     seacr_top = 0.01))

```
