---
title: "Processing Sequencing data with pipelineNGS"
author: "Mireia Ramos-Rodriguez"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    fig.align: "center"
    self.contained: TRUE
vignette: >
  %\VignetteIndexEntry{Analyzing UMI-4C data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

```{r Setup, message=FALSE, warning=FALSE, echo=FALSE}
devtools::load_all()
knitr::opts_chunk$set(cache=TRUE)
.dat <- "test/"
```

# Requirements

## Software

-   [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
-   [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
-   [Samtools](http://www.htslib.org/)
-   [bedtools](http://bedtools.readthedocs.io/en/latest/). Necessary for ATAC-seq offset correction and SEACR peak calling.
-   [MACS2](https://github.com/taoliu/MACS)

## External files

Optional:

-   ENCODE Blacklist (`blacklist`). You can download a bed file with the hg19 blacklisted regions from UCSC's [[wgEncodeDacMapabilityConsensusExcludable.bed.gz](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/)].

Mandatory for ATAC-seq offset correction and SEACR peak calling:

-   Chromosome sizes (`gen_sizes`). You can generate this file by using `fetchChromSizes` from [UCSC tools](http://hgdownload.soe.ucsc.edu/admin/exe/) or downloading int directly from the UCSC: [hg38](https://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes) or [hg19](https://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes).

## Recommended directory organization

I recommend you create a data directory within your project with the following structure: - Data directory: `data/` - 1 folder for each type of experiment: `ATAC-seq`; `ChiP-seq`. - `BAM/` output BAM files from alignment and post-processing. - `Logs/` directory where the log outputs from different programs are saved. - `Visualization/` directory to output the visualization files (bigWig and bedGraphs). - `Peaks/` directory with `.narrowPeak`, `.broadPeak` and other files created by the peak calling analysis.

# Pipeline Overview

For the purpose of this vignette we are going to focus on analyzing ATAC-seq and ChIP-seq data because, in general, the steps needed for processing this data are the same.

In the following figure you can see a description of the steps needed for the analysis of each type of experiment, and the functions provided in this package that perform each one of the steps.

```{r, echo=FALSE}
knitr::include_graphics("figures/align_postpro.png", dpi=100)
```

## Quick example reference

> The input names for the pipeline should be the name of the samples without any suffix. A suffix specific for each analysis and type of file will be added in each step of the analysis.

```{r}
##-------------------------------------------------------
## Input data
##-------------------------------------------------------
# Input sample names (without any suffix)
samples <- c("ctrl-2", "mut-2", "mut-3", "mut-4")

# Input folders
path_bam <- "BAM/"
path_fastq <- "raw/"
logs <- "logs/"
vis <- "visualization/"
pc <- "peak_calling_shift/"

##-------------------------------------------------------
## Run functions
##-------------------------------------------------------
# Load alignment library
library(pipelineNGS)

# Alignment
alignmentBowtie2(samples, path_bam=path_bam, path_fastq=path_fastq, 
                 suffix_fastq=".fastq.gz", path_logs=logs)
 
# Post-processing ATAC-seq
samToBam(samples, path_bam=path_bam, path_logs=logs)
sortAndStats(samples, path_bam=path_bam, path_logs=logs,
            input_suf=".uns.raw.bam", output_suf=".raw.bam")
filterOutReads(samples, path_bam=path_bam, path_logs=logs)
removeDuplicates(samples, path_bam=path_bam, path_logs=logs)
offsetATAC(samples, path_bam=path_bam, path_logs=logs)
sortAndStats(samples, path_bam=path_bam, path_logs=logs,
             input_suf=".rmdup.bam", output_suf=".noOffset.bam")
summarizeLogs(out_name="summary", path_logs=logs)

# Create bigWig
bamToVis(samples=samples, summaryStats=paste0(logs, "summary.rda"),
       path_bam=path_bam, path_vis=vis)

# Peak calling
callPeak(samples, path_bam=path_bam,
        path_pc=pc, path_logs=logs,
        type="narrow")
```

## Quality control with FASTQC

The first step for any NGS pipeline is checking quality of the reads. This can be done using a program called [**FastQC**](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), which takes as input FASTQ files and returns an html with several quality control analysis and if your sequences passed/failed each test.

We can use the function `fastqc()` to perform this analysis in one or several samples in a row.

```{r fastqc, eval=FALSE, include=TRUE}
fastqc(samples="thyroid_sample", out_dir="testData/fastqc/", path_fastq="testData/raw/")
```

## Aligning data

For the alignment of ATAC and ChIP-seq data, we are going to use [**bowtie2**](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) as an aligner. The output will be a SAM file. The parameters with which we are running the program are as follows:

-   `-t`: Print wall-clock time taken by search phases.
-   `-x <bt2-idx>`: Index filename prefix (minus trailing .X.bt2). In our case it is already pointed to `/biodata/`, where the index is stored.[^1].
-   `-U`/`-1`,`-2`: Indicates the location and names of the fastq files to align. `-U` stands for single-end reads and `-1` and `-2` for the pair 1 reads and pair 2 reads, respectively (paired-end reads).
-   `-p`: Number of alignment threads to launch.

[^1]: See [Bowtie2-build indexer](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer) for more info on how the index is created.

The function we need to use for performing the alignment for this kind of data is `alignmentBowtie2()`. The default index is set as hg19 (`/biodata/indices/species/Hsapiens/hg19`)

```{r alignmentBowtie2, eval=FALSE, include=TRUE}
alignmentBowtie2(samples="thyroid_sample", type="SE", 
                 path_fastq="testData/raw/", suffix_fastq=".fq.gz", path_bam="testData/bam/", 
                 cores=6, path_logs="testData/logs/")
```

You can check the statistics of the alignment in the `logs/` folder and the result for the aligment can be found in `bam/tmp/`.

```{bash cat alignmentBowtie2}
cat testData/logs/thyroid_sample.alignment.log
```

<!-- ### RNA-seq -->

<!-- When aligning RNA-seq data, we need to use a different alignment. In our case, we chose to use [TopHat2](https://ccb.jhu.edu/software/tophat/index.shtml) -->

<!-- - `-r`: Mate inner distance (set at 100, default tophat 50). -->

<!-- - `-o`: Output directory for the results. -->

<!-- - `-p`: Number of alignment threads to launch. -->

<!-- - `--no-coverage-search`: -->

<!-- - `--transcriptome-index`:  -->

## Data post-processing

After the aligment, we need to do several post-processing steps in order to have homogeneous and filtered data for the downstream analysis. This post-processing is very similar between ATAC-seq and ChIP-seq, but ATAC-seq needs an extra step for filtering the reads.

First of all, since the output from **bowtie2** is a `SAM` file, we need to **convert** it into its **binary-compressed form: `BAM`**. Therefore, the first function to use is `samToBam()`, which will convert the SAM file to BAM (`samtools view`). Then, we'll also need to run `sortAndStats()` to sort it by coordinates (`samtools sort`), create an index (`samtools index`) and a summary (`samtools flagstat`).

```{r samToBam, include=TRUE, eval=FALSE}
samToBam(samples="thyroid_sample", path_bam="testData/bam/", path_logs="testData/logs/")
sortAndStats(samples="thyroid_sample", path_bam="testData/bam/", path_logs="testData/logs/",
             input_suf=".uns.raw.bam", output_suf=".raw.bam")
```

```{bash cat samToBam}
cat testData/logs/thyroid_sample.raw.stats
```

As we can observe in the summary report, this file still contains not aligned sequences. Therefore, the next step is removing those reads that are not aligned or aligned in regions that are not interesting for us. The function to perform this filtering is `filterOutReads()`. You can provide the specific chromosomes you want to filter out by using the argument `remove=`. By default it removes non-chanonical chromosomes. It will also automatically remove the reads aligned to the ENDOCE Consensus Blacklist regions[^2].

[^2]: [wgEncodeDacMapabilityConsensusExcludable.bed.gz](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/)

```{r filterOutReads, include=TRUE, eval=FALSE}
filterOutReads(samples="thyroid_sample", path_bam="testData/bam/", path_logs="testData/logs/")
```

```{bash cat filterOutReads}
cat testData/logs/thyroid_sample.filtered.stats
```

Once we have removed the reads that are not informative for us, we also need to **remove** those reads that could be **PCR duplicates**. To do so, we use [**Picard Tools**](https://broadinstitute.github.io/picard/) implemented in the function `removeDuplicates()`.

```{r removeDuplicates, include=TRUE, eval=FALSE}
removeDuplicates(samples="thyroid_sample", path_bam="testData/bam/", path_logs="testData/logs/")
```

```{bash cat removeDuplicates}
tail -n 4 testData/logs/thyroid_sample.rmdup.stats
```

If we are dealing with **ATAC-seq** data, we need to perform an **offset correction**. This is implemented in the function `offsetATAC()`, which uses [**bedtools**](http://bedtools.readthedocs.io/en/latest/) and `awk` to correct the reads for the transposase binding event.

> **Warning**: If you are trying to do some kind of allelle specific analysis or any other analysis in which it is important the actual sequence of the read and its location, you might have to use BAM files without the offset correction.

```{r Offset Correction, include=TRUE, eval=FALSE}
offsetATAC(samples="thyroid_sample", path_bam="testData/bam/",path_logs="testData/logs/")
```

Since the `offsetATAC()` function already sorts and indexes the file, we do not need to run any further functions.

However, if you are analyzing **ChIP-seq data** or you don't want to perform an offset correction for your ATAC-seq data, you'll need to run the function `sortAndStats()` with default parameters, in order to create the final file and add the statistics to the log folder.

Once the post-processing is completed, you need to check if the final files are correct and then you can proceed to remove the temporary folder where all intermidiate files are stored using `rmTmp()`.

> Do not remove the tmp folder before checking that the files are correct, because if they are not, you'll need to run the whole pipeline again.

```{r Remove tmp, include=TRUE, eval=FALSE}
rmTmp(path_bam="testData/bam/")
```

Before proceding with the analysis of the sequences, we can generate a summary table with the number of reads that were filtered at each step using `summarizeLogs()`. The different values that are stored in this table are:

-   `al.raw`: Total number of reads obtained from the sequencing facility.
-   `al.aligned`: Total number of aligned reads. $$al.aligned=al.raw−al.unaligned$$
-   `pp.clean`: Number of aligned reads after removing those aligned to chrM and non-cannonical chromosomes. $$pp.clean=al.aligned−pp.filtOut$$
-   `pp.final`: Final number of reads, after removing duplicated reads. $$pp.final=pp.clean−pp.duplicates$$
-   `perc.align`: Percentage of aligned reads from total raw reads obtained. $$perc.align=al.alignedal.raw∗100$$
-   `perc.filtOut`: Percentage of the aligned reads that are removed because they map to non-cannonical chromosomes. $$perc.filtOut=pp.filtOutal.aligned∗100$$
-   `perc.duplicates`: Percentage of the aligned reads that are marked as duplicates, and therefore removed. $$perc.duplicates=pp.duplicates - al.aligned∗100$$
-   `perc.effiRaw`: Percentage of the raw reads that are kept from dowsntream analysis (after removing duplicates). $$perc.effiRaw=pp.finalal.raw∗100$$
-   `perc.effiAligned`: Percentage from the raw reads that are kept from dowsntream analysis (after removing duplicates). $$perc.effiAligned=pp.finalal.aligned∗100$$
-   `scale`: Factor to use for scaling track to RPKM (10K reads). It is used by function `bamToVis()`. $$scale=\frac{1e^7}{pp.clean}$$

```{r Summary Stats}
summarizeLogs(out_name="summary", path_logs="testData/logs/")
load("testData/logs/summary.rda")
knitr::kable(sum.df)
```

# Conversion to bigWig and Visualization

Once we have our post-processed data, we can proceed to visualize the coverage of our samples in a genome browser. We can choose between uploading the tracks to public servers and load them into [**UCSC Genome Browser**](https://genome.ucsc.edu/cgi-bin/hgTracks) or to use a local browser like [**IGV**](http://software.broadinstitute.org/software/igv/).

IGV allows to load BAM files and see the coverage and alignment of the reads, but if we want to look just at the distribution of reads, we need to create either BedGraph of BigWig files. In oder to be inclusive, we generated a function to create both BedGraph-compressed (.tar.gz) and BigWig files.

```{r Visualization, eval=FALSE, message=TRUE, warning=FALSE, include=TRUE}
bamToVis(samples="thyroid_sample", summaryStats="testData/logs/summary.rda",
         path_bam="testData/bam/", path_vis="testData/vis/")
```

Now we can load these files to IGV or UCSC to see the profiles.

## Peak calling

In order to proceed with the analysis it is also interesting to have the coordinates for the complete set of enriched regions in our experiment. The process of searching and reporting such regions is the **peak calling**.

To run these analysis we are going to use [**MACS2**](https://github.com/taoliu/MACS) callPeak function, which is implemented in `callPeak()`. We should include the parameter `type` which is `"narrow"` as default. For ATAC-seq experiments and ChIP-seq for TF we need to use `type="narrow"` (default) and for ChIP-seq for histone marks we need to use `type="broad"`.

> For running this in IGTP machines, you first need to activate the virtual environment for the last MACS2 version. This means that before running your Rscript with such functions in your terminal, youll need to run `activate-macs-git-2017.5.15`.

This function will return several files in the folder `path_pc` corresponding to our enriched regions.

```{r Peak calling, include=TRUE, eval=FALSE}
callPeak("thyroid_sample", path_bam="testData/bam/", 
         path_pc="testData/peakCalling/", path_logs="testData/logs/", 
         type="narrow")
```
