<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This package provides functions for processing FASTQ files and doing the downstream analysis, such as peak calling, conversion to bigWig format or motif analysis.">
<title>Methods for Epigenetic Data Processing â€¢ pipelineNGS</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Methods for Epigenetic Data Processing">
<meta property="og:description" content="This package provides functions for processing FASTQ files and doing the downstream analysis, such as peak calling, conversion to bigWig format or motif analysis.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">pipelineNGS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/introduction.html">Processing epigenome data with pipelineNGS</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="pipelinengs">pipelineNGS<a class="anchor" aria-label="anchor" href="#pipelinengs"></a>
</h1></div>
<p><code>pipelineNGS</code> is a package for processing epigenomic high-throughput data, specifically histone mark ChIP-seq and ATAC-seq.</p>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="programs-you-need-to-have-installed-and-included-in-your-path">Programs you need to have installed and included in your path:<a class="anchor" aria-label="anchor" href="#programs-you-need-to-have-installed-and-included-in-your-path"></a>
</h3>
<p>As this package is a wrapper for some command line tools, you need to have this programs in your <code>$PATH</code>. If they are not in your <code>$PATH</code>, you can also provide the path to the binary files using the appropriate arguments.</p>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" class="external-link">bowtie2</a></li>
<li><a href="http://www.htslib.org/" class="external-link">Samtools</a></li>
<li><a href="https://github.com/taoliu/MACS" class="external-link">MACS2</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="files-you-need-to-have-in-your-local-machine">Files you need to have in your local machine:<a class="anchor" aria-label="anchor" href="#files-you-need-to-have-in-your-local-machine"></a>
</h3>
<p>Additionally, you will need to download reference files to perform the different steps in the pre-processing pipeline:</p>
<ul>
<li>Reference genome indexed with Bowtie2. See more information <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#indexing-a-reference-genome" class="external-link">here</a>. You can download the fasta files for the different genome builds from the <a href="https://hgdownload.soe.ucsc.edu/downloads.html#human" class="external-link">UCSC download site</a>.</li>
<li>Chromosome sizes (<code>gen_sizes</code>). You can download this file also from the UCSC: <a href="https://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes" class="external-link">hg38</a> or <a href="https://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes" class="external-link">hg19</a>. Only necessary for ATAC-seq offset correction.</li>
<li>ENCODE Blacklist (<code>blacklist</code>). They can be downloaded from <a href="https://sites.google.com/site/anshulkundaje/projects/blacklists" class="external-link">here</a>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="installing-pipelinengs-in-your-local-machine">Installing pipelineNGS in your local machine<a class="anchor" aria-label="anchor" href="#installing-pipelinengs-in-your-local-machine"></a>
</h2>
<p>Open your R session, install the <code>devtools</code> package if it is not already in your machine and type the following:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install pipelineNGS package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"mireia-bioinfo/pipelineNGS"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pipelineNGS package</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pipelineNGS)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pipeline-overview">Pipeline Overview<a class="anchor" aria-label="anchor" href="#pipeline-overview"></a>
</h2>
<p>In this package we currently have implemented the pipelines for analyzing the following experiments:</p>
<ul>
<li>ATAC-seq (<code>ATAC</code>).</li>
<li>ChIP-seq for histone marks (<code>CHIP</code>).</li>
<li>CUTandTAG for histone marks (<code>CT</code>).</li>
<li>CUTandRUN for transcription factors (<code>CR</code>).</li>
</ul>
<p>In the following figure you can see a description of the steps needed for the analysis of each type of experiment, with specific arguments (if any) used in the different steps.</p>
<p><img src="articles/figures/process_epi_map.png"></p>
<p>Here is an example on how to run a ChIP-seq analysis with single-end data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## General parameters</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="st">"/vault/refs/indexes/hg38"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>blacklist <span class="ot">&lt;-</span> <span class="st">"/vault/refs/Blacklist/lists/hg38-blacklist.v2.bed"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Example Single End ##</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>fastq_files <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fastq/sample1_L1.fastq.gz"</span>, <span class="st">"fastq/sample1_L2.fastq.gz"</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"fastq/sample1_L3.fastq.gz"</span>, <span class="st">"fastq/sample2_L2.fastq.gz"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"fastq/sample3_L1.fastq.gz"</span>, <span class="st">"fastq/sample3_L3.fastq.gz"</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Convert to list to use as input for process_epigenome()</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create one list element for each simple</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">basename</span>(fastq_files), <span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>])</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>fastq_input <span class="ot">&lt;-</span> <span class="fu">split</span>(fastq_files, names)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>fastq_input</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="do">## Using the files described in the previous chunk:</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="fu">process_epigenome</span>(<span class="at">fastq_files=</span>fastq_input,</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">out_name=</span><span class="fu">names</span>(fastq_input),</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>                  <span class="at">run_fastqc=</span><span class="cn">TRUE</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>                  <span class="at">seq_type=</span><span class="st">"CT"</span>,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                  <span class="at">type=</span><span class="st">"PE"</span>,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                  <span class="at">index=</span>index,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                  <span class="at">blacklist=</span>blacklist,</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>                  <span class="at">cores=</span><span class="dv">6</span>)</span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pipelineNGS</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Mireia Ramos-Rodriguez <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
