<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pipelineNGS">
<title>CUT&amp;TAG Processing Pipeline • pipelineNGS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CUT&amp;TAG Processing Pipeline">
<meta property="og:description" content="pipelineNGS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pipelineNGS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CUT&amp;TAG_processing_pipeline.html">CUT&amp;TAG Processing Pipeline</a>
    <a class="dropdown-item" href="../articles/introduction.html">Processing Sequencing data with pipelineNGS</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>CUT&amp;TAG Processing Pipeline</h1>
                        <h4 data-toc-skip class="author">Mireia
Ramos-Rodríguez</h4>
            <address class="author_afil">
      Department of Experimental and Health Sciences, University Pompeu
Fabra<br><a class="author_email" href="mailto:#"></a><a href="mailto:mireia.ramos@upf.edu" class="email">mireia.ramos@upf.edu</a>
      </address>
                  
      
      
      <div class="d-none name"><code>CUT&amp;TAG_processing_pipeline.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes the pipeline implemented to analyze
      CUT&amp;TAG data.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The mapping of chromatin features genome-wide has traditionally been
performed using chromatin immunoprecipitation (ChIP), in which chromatin
is cross-linked and solubilized, and an antibody to a protein or
modification of interest is used to immunoprecipitate the bound DNA
(Fig. 1a).</p>
<p>An alternative chromatin profiling strategy is enzyme tethering in
situ whereby the chromatin protein or modification of interest is
targeted by an antibody or fusion protein. Then, the underlying DNA is
marked or cleaved, and a succession of enzyme-tethering methods have
been introduced over the past two decades. Cleavage Under Targets &amp;
Tagmentation (CUT&amp;Tag) is a tethering method that uses a
protein-A-Tn5 (pA-Tn5) transposome fusion protein (Fig. 1b).</p>
<div class="figure">
<img src="https://yezhengstat.github.io/CUTTag_tutorial/figures/ChIPseqCUTTag.png" alt=""><p class="caption">Differences between immunoprecipitation and in
antibody-targeted chromatin profiling strategies. A. ChIP-seq
experimental procedure. B. CUT&amp;Tag experimental procedure. Cells and
nuclei are indicated in grey, chromatin as red nucleosomes, and a
specific chromatin protein in green.</p>
</div>
<p>This package implements some functions to facilitate processing of
CYT&amp;TAG data, based on the original published CUT&amp;TAG analysis
tutorial.</p>
<blockquote>
<p>Original <strong>CUT&amp;TAG pipeline</strong>: <a href="https://yezhengstat.github.io/CUTTag_tutorial/" class="external-link uri">https://yezhengstat.github.io/CUTTag_tutorial/</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h2>
<ul>
<li>
<strong>Bowtie2</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" class="external-link uri"&gt;http://bowtie-bio.sourceforge.net/bowtie2/index.shtml&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> (version 2.4.1).</li>
<li>
<strong>Samtools</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="http://www.htslib.org/" class="external-link uri"&gt;http://www.htslib.org/&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a> (version 1.10).</li>
<li>
<strong>Bedtools</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://bedtools.readthedocs.io/en/latest/" class="external-link uri"&gt;https://bedtools.readthedocs.io/en/latest/&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a> (version 2.29.2).</li>
<li>
<strong>SEACR</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://github.com/FredHutch/SEACR" class="external-link uri"&gt;https://github.com/FredHutch/SEACR&lt;/a&gt;&lt;/p&gt;'><sup>4</sup></a> (version 1.3).</li>
</ul>
</div>
<div class="section level2">
<h2 id="pipeline-description">Pipeline description<a class="anchor" aria-label="anchor" href="#pipeline-description"></a>
</h2>
<div class="section level3">
<h3 id="fastq-quality-control-with-fastqc">FastQ Quality Control with FastQC<a class="anchor" aria-label="anchor" href="#fastq-quality-control-with-fastqc"></a>
</h3>
<pre><code>fastqc -t 5 -o FastQC/ /path/to/fastq/sample_1_read1.fastq.gz</code></pre>
<p>The <strong>FAIL due to discordant sequence content</strong> at the
beginning of the reads are a <strong>common phenomenon for CUT&amp;Tag
reads</strong> and is likely related to a Tn5 sequence preference.
Authors do not recommend read trimming, as the bowtie2 parameters that
they list will give accurate mapping information without trimming.</p>
</div>
<div class="section level3">
<h3 id="aligment-with-bowtie2">Aligment with Bowtie2<a class="anchor" aria-label="anchor" href="#aligment-with-bowtie2"></a>
</h3>
<pre><code>bowtie2 -t -x ~/refs/indexes/hg38 -1 /path/to/fastq/sample_1_read1.fastq.gz -2 /path/to/fastq/sample_1_read2.fastq.gz -p 5 2&gt; Logs/sample_1.alignment.log | samtools view - -b -@ 4 | samtools sort - -@ 4 -m 2G -o BAM/sample_1.raw.bam ; samtools index BAM/sample_1.raw.bam -@ 4 ; samtools idxstats BAM/sample_1.raw.bam &gt; Logs//sample_1.raw.idxstats.log</code></pre>
<p>This command is optimized for obtaining very sensitive alignments
without the need of trimming adapters from the input reads.</p>
</div>
<div class="section level3">
<h3 id="filtering-out-unnecessary-reads">Filtering out unnecessary reads<a class="anchor" aria-label="anchor" href="#filtering-out-unnecessary-reads"></a>
</h3>
<p>Remove non-cannonical, unmapped and blacklisted reads:</p>
<pre><code>samtools idxstats BAM/10759_1.raw.bam | cut -f 1 | grep -v chrM | grep -v chrUn | grep -v _random | grep -v _hap | grep -v _gl | grep -v EBV | xargs samtools view 
-b BAM/10759_1.raw.bam -@ 4 -F 4 | samtools view - -b -L ~/refs/Blacklist/lists/hg38-blacklist.v2.bed -U BAM/10759_1.tmp.bam -o trash.bam </code></pre>
<p>Fixmates:</p>
<pre><code>samtools sort -n BAM/10759_1.tmp.bam -@ 4 -m 2G -o - | samtools fixmate -m - - | samtools sort - -@ 4 -m 2G -o BAM/10759_1.fixmate.bam </code></pre>
<p>Remove duplicates:</p>
<pre><code>samtools markdup BAM/10759_1.fixmate.bam BAM/10759_1.bam -r -s 2&gt; Logs/10759_1.rmdup.log ; samtools index BAM/10759_1.bam -@ 4 ; samtools idxstats BAM/10759_1.bam </code></pre>
</div>
<div class="section level3">
<h3 id="peak-calling-with-seacr">Peak calling with SEACR<a class="anchor" aria-label="anchor" href="#peak-calling-with-seacr"></a>
</h3>
<p>Convert to appropriate format:</p>
<pre><code># Convert to BED
samtools sort -n BAM/sample_1.bam -@ 4-m 2G -o - | bedtools bamtobed -bedpe -i stdin | awk '$1==$4 &amp;&amp; $6-$2 &lt; 1000 {print $0}' | cut -f 1,2,6 | sort -dsk1,1 -k2n,2 -k3nr,3 &gt; BAM/sample_1.bed
# Convert to BedGraph
bedtools genomecov -bg -i sample_1.bed -g ~/refs/hg38.chrom.sizes &gt; sample_1.bedgraph</code></pre>
<p>Peak calling:</p>
<pre><code>SEACR_1.3.sh BAM/sample_1.bam 0.01 norm stringent Peaks/sample_1.peaks.0.01</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="running-the-pipeline">Running the pipeline<a class="anchor" aria-label="anchor" href="#running-the-pipeline"></a>
</h2>
<ul>
<li>For paired-end samples, the input <code>fastq_files</code> should be
a list of length equal to the number of samples, with each element
containing 2 file paths (read 1 and read 2).</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/sample_info.tsv"</span><span class="op">)</span><span class="op">)</span>
<span class="va">fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">files</span>, <span class="va">samples</span><span class="op">$</span><span class="va">sampleID</span><span class="op">)</span> <span class="co"># Create a list with each sample containing the two read files (R1 and R2)</span></code></pre></div>
<div class="section level3">
<h3 id="each-step-for-all-samples">Each step for all samples<a class="anchor" aria-label="anchor" href="#each-step-for-all-samples"></a>
</h3>
<p>By default, the function <code>processCUTTAG</code> runs each step of
the pipeline for each sample, before jumping to the next step.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/processCUTTAG.html">processCUTTAG</a></span><span class="op">(</span>fastq_files<span class="op">=</span><span class="va">fastq</span>,
              out_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fastq</span><span class="op">)</span>,
              MoreArgs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>index<span class="op">=</span><span class="st">"~/refs/indexes/hg38"</span>,
                            blacklist<span class="op">=</span><span class="st">"~/refs/Blacklist/lists/hg38-blacklist.v2.bed"</span>,
                            chr_sizes<span class="op">=</span><span class="st">"~/refs/hg38.chromSizes.txt"</span>,
                            cores<span class="op">=</span><span class="fl">5</span>,
                            seacr_type <span class="op">=</span> <span class="st">"stringent"</span>,
                            seacr_top <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="one-full-sample-at-a-time">One full sample at a time<a class="anchor" aria-label="anchor" href="#one-full-sample-at-a-time"></a>
</h3>
<p>In case you want to run all the pipeline one sample at a time, you
can take advantage of the <code>mapply</code> function, which will run
<code>processCUTTAG</code> for each element of the provided list.`</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">processCUTTAG</span>,
       fastq_files<span class="op">=</span><span class="va">fastq</span>,
       out_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fastq</span><span class="op">)</span>,
       MoreArgs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>index<span class="op">=</span><span class="st">"~/refs/indexes/hg38"</span>,
                     blacklist<span class="op">=</span><span class="st">""</span>, <span class="co"># Do not remove blacklisted reads</span>
                     chr_sizes<span class="op">=</span><span class="st">"~/refs/hg38.chromSizes.txt"</span>,
                     cores<span class="op">=</span><span class="fl">5</span>,
                     seacr_type <span class="op">=</span> <span class="st">"stringent"</span>,
                     seacr_top <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
