<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pipelineNGS">
<title>Processing epigenome data with pipelineNGS • pipelineNGS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Processing epigenome data with pipelineNGS">
<meta property="og:description" content="pipelineNGS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pipelineNGS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction.html">Processing epigenome data with pipelineNGS</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Processing epigenome data with pipelineNGS</h1>
                        <h4 data-toc-skip class="author">Mireia
Ramos-Rodriguez</h4>
            
      
      
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Package installation:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"mireia-bioinfo/pipelineNGS"</span><span class="op">)</span></code></pre></div>
<p>Load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pipelineNGS</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level3">
<h3 id="software">Software<a class="anchor" aria-label="anchor" href="#software"></a>
</h3>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" class="external-link">bowtie2</a></li>
<li><a href="http://www.htslib.org/" class="external-link">Samtools</a></li>
<li>
<a href="http://bedtools.readthedocs.io/en/latest/" class="external-link">bedtools</a>.
Necessary for ATAC-seq offset correction and SEACR peak calling.</li>
<li><a href="https://github.com/taoliu/MACS" class="external-link">MACS2</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="external-files">External files<a class="anchor" aria-label="anchor" href="#external-files"></a>
</h3>
<p>Optional:</p>
<ul>
<li>ENCODE Blacklist (<code>blacklist</code>). You can download a bed
file with the hg19 blacklisted regions from UCSC’s [<a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/" class="external-link">wgEncodeDacMapabilityConsensusExcludable.bed.gz</a>].</li>
</ul>
<p>Mandatory for ATAC-seq offset correction and SEACR peak calling:</p>
<ul>
<li>Chromosome sizes (<code>gen_sizes</code>). You can generate this
file by using <code>fetchChromSizes</code> from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/" class="external-link">UCSC tools</a> or
downloading it directly from the UCSC: <a href="https://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes" class="external-link">hg38</a>
or <a href="https://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes" class="external-link">hg19</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="recommended-directory-organization">Recommended directory organization<a class="anchor" aria-label="anchor" href="#recommended-directory-organization"></a>
</h3>
<p>I recommend you create a data directory within your project with the
following structure:</p>
<ul>
<li>Data directory: <code>data/</code>
<ul>
<li>1 folder for each type of experiment: <code>ATAC-seq</code>;
<code>ChiP-seq</code>.
<ul>
<li>
<code>BAM/</code> output BAM files from alignment and
post-processing.</li>
<li>
<code>Peaks/</code> directory with <code>.narrowPeak</code>,
<code>.broadPeak</code> and other files created by the peak calling
analysis.</li>
<li>
<code>Logs/</code> directory where the log outputs from different
programs are saved.</li>
<li>
<code>Visualization/</code> directory to output the visualization
files (bigWig and bedGraphs).</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="pipeline-overview">Pipeline Overview<a class="anchor" aria-label="anchor" href="#pipeline-overview"></a>
</h2>
<p>In this package we currently have implemented the pipelines for
analyzing the following experiments:</p>
<ul>
<li>ATAC-seq (<code>ATAC</code>).</li>
<li>ChIP-seq for histone marks (<code>CHIP</code>).</li>
<li>CUTandTAG for histone marks (<code>CT</code>).</li>
<li>CUTandRUN for transcription factors (<code>CR</code>).</li>
</ul>
<p>In the following figure you can see a description of the steps needed
for the analysis of each type of experiment, with specific arguments (if
any) used in the different steps.</p>
<p><img src="figures/process_epi_map.png" width="95%"></p>
<p>The specific parameters used for each experiment are described in
<code>paramdata</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">paramdata</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="9%">
<col width="54%">
<col width="14%">
<col width="8%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th align="left">Parameter</th>
<th align="left">Experiment</th>
<th align="left">Extra_bowtie2</th>
<th align="left">Offset_correction</th>
<th align="left">peak_type</th>
<th align="left">shift</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATAC</td>
<td align="left">ATAC-seq</td>
<td align="left"></td>
<td align="left">TRUE</td>
<td align="left">narrow</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">CHIP</td>
<td align="left">ChIP-seq</td>
<td align="left"></td>
<td align="left">FALSE</td>
<td align="left">broad</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">CT</td>
<td align="left">CUT&amp;TAG</td>
<td align="left">–very-sensitive –no-mixed –no-discordant –phred33 -I 10
-X 700</td>
<td align="left">FALSE</td>
<td align="left">broad</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">CR</td>
<td align="left">CUT&amp;RUN</td>
<td align="left"></td>
<td align="left">FALSE</td>
<td align="left">narrow</td>
<td align="left">FALSE</td>
</tr>
</tbody>
</table>
<p>All these steps are implemented in a single function called
<code><a href="../reference/process_epigenome.html">process_epigenome()</a></code>. Critical parameters for this function
are the following:</p>
<ul>
<li>
<code>fastq_files</code>. In this argument you must provide as a
character vector (only when reads are single end and each file
corresponds to a single sample) or a list the complete paths for the
fastq files you wish to process. A list input is mandatory for
paired-end samples (each element of the list contains paths to R1 and
R2), however you can use lists for single end reads if you have several
files for a single sample that has been sequenced in different
lines:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Example Paired End ##</span>
<span class="va">fastq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fastq/sample1_L1_R1.fastq.gz"</span>, <span class="st">"fastq/sample1_L2_R1.fastq.gz"</span>,
                 <span class="st">"fastq/sample1_L1_R2.fastq.gz"</span>, <span class="st">"fastq/sample1_L2_R2.fastq.gz"</span>,
                 <span class="st">"fastq/sample2_L1_R1.fastq.gz"</span>, <span class="st">"fastq/sample2_L1_R2.fastq.gz"</span><span class="op">)</span>

<span class="co">## Convert to list to use as input for process_epigenome()</span>
<span class="co"># Create one list element for each simple</span>
<span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fastq_files</span><span class="op">)</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">fastq_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">fastq_files</span>, <span class="va">names</span><span class="op">)</span>

<span class="co"># Make lists with R1 and R2 inside each sample</span>
<span class="va">fastq_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fastq_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>R1<span class="op">=</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"_R1.fastq.gz"</span>, <span class="va">x</span><span class="op">)</span><span class="op">]</span>,
                                    R2<span class="op">=</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"_R2.fastq.gz"</span>, <span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">fastq_input</span></code></pre></div>
<pre><code><span class="co">## $sample1</span>
<span class="co">## $sample1$R1</span>
<span class="co">## [1] "fastq/sample1_L1_R1.fastq.gz" "fastq/sample1_L2_R1.fastq.gz"</span>
<span class="co">## </span>
<span class="co">## $sample1$R2</span>
<span class="co">## [1] "fastq/sample1_L1_R2.fastq.gz" "fastq/sample1_L2_R2.fastq.gz"</span>
<span class="co">## </span>
<span class="co">## </span>
<span class="co">## $sample2</span>
<span class="co">## $sample2$R1</span>
<span class="co">## [1] "fastq/sample2_L1_R1.fastq.gz"</span>
<span class="co">## </span>
<span class="co">## $sample2$R2</span>
<span class="co">## [1] "fastq/sample2_L1_R2.fastq.gz"</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Example Single End ##</span>
<span class="va">fastq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fastq/sample1_L1.fastq.gz"</span>, <span class="st">"fastq/sample1_L2.fastq.gz"</span>,
                 <span class="st">"fastq/sample1_L3.fastq.gz"</span>, <span class="st">"fastq/sample2_L2.fastq.gz"</span>,
                 <span class="st">"fastq/sample3_L1.fastq.gz"</span>, <span class="st">"fastq/sample3_L3.fastq.gz"</span><span class="op">)</span>

<span class="co">## Convert to list to use as input for process_epigenome()</span>
<span class="co"># Create one list element for each simple</span>
<span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fastq_files</span><span class="op">)</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">fastq_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">fastq_files</span>, <span class="va">names</span><span class="op">)</span>
<span class="va">fastq_input</span></code></pre></div>
<pre><code><span class="co">## $sample1</span>
<span class="co">## [1] "fastq/sample1_L1.fastq.gz" "fastq/sample1_L2.fastq.gz"</span>
<span class="co">## [3] "fastq/sample1_L3.fastq.gz"</span>
<span class="co">## </span>
<span class="co">## $sample2</span>
<span class="co">## [1] "fastq/sample2_L2.fastq.gz"</span>
<span class="co">## </span>
<span class="co">## $sample3</span>
<span class="co">## [1] "fastq/sample3_L1.fastq.gz" "fastq/sample3_L3.fastq.gz"</span></code></pre>
<ul>
<li>
<code>out_name</code>. Names to use for the output files. Should be
the same length as the list/vector used as input in
<code>fastq_files</code>. The input names for the pipeline should be the
name of the samples without any suffix. A suffix specific for each
analysis and type of file will be added in each step of the
analysis.</li>
<li>
<code>seq_type</code>. The type of experiment you are processing.
Available options are listed at the beginning of this section.</li>
<li>
<code>type</code>. Either <code>SE</code> for single-end reads or
<code>PE</code> for paired-end reads.</li>
</ul>
<p>Additional arguments refer to specific parts of the processing and
will be described in the appropriate sections.</p>
</div>
<div class="section level2">
<h2 id="data-processing-steps">Data processing steps<a class="anchor" aria-label="anchor" href="#data-processing-steps"></a>
</h2>
<p>All needed data processing steps are wrapped inside the function
<code>process_epigenome</code>. We will now go through every key step to
explain the important parameters.</p>
<p>The following chunk shows a minor example of how to run an analyses,
in this case of paired-end (<code>type=PE</code>) CUTandTAG data
(<code>seq_type="CT"</code>):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">index</span> <span class="op">&lt;-</span> <span class="st">"/vault/refs/indexes/hg38"</span>
<span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="st">"/vault/refs/Blacklist/lists/hg38-blacklist.v2.bed"</span>

<span class="co"># Using the files described in the previous chunk:</span>
<span class="fu"><a href="../reference/process_epigenome.html">process_epigenome</a></span><span class="op">(</span>fastq_files<span class="op">=</span><span class="va">fastq_input</span>,
                  out_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fastq_input</span><span class="op">)</span>,
                  run_fastqc<span class="op">=</span><span class="cn">TRUE</span>,
                  seq_type<span class="op">=</span><span class="st">"CT"</span>,
                  type<span class="op">=</span><span class="st">"PE"</span>,
                  index<span class="op">=</span><span class="va">index</span>,
                  blacklist<span class="op">=</span><span class="va">blacklist</span>,
                  cores<span class="op">=</span><span class="fl">6</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>The first step for any NGS pipeline is checking quality of the reads.
This can be done using a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link"><strong>FastQC</strong></a>,
which takes as input FASTQ files and returns an html with several
quality control analysis and if your sequences passed/failed each
test.</p>
<p>This part is implemented inside <code><a href="../reference/process_epigenome.html">process_epigenome()</a></code> and
will run when argument <code>run_fastqc=TRUE</code>. You can use the
function <code><a href="../reference/fastqc.html">fastqc()</a></code> to perform only this analysis.</p>
<p>Results from this step are stored in a folder named
<code>FastQC/</code>.</p>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>For the alignment of epigenetic data we use <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" class="external-link"><strong>bowtie2</strong></a>
as an aligner. The output will be redirected into a pipe to create a
position-sorted and indexed BAM file (<code>samplename.raw.bam</code>).
The parameters used to run bowtie2 are:</p>
<ul>
<li>
<code>--local</code>: Perform local alignment (will clip sequences
that do not match the reference). This allows us to align without
removing the adapters, as they will be clipped from the sequence.</li>
<li>
<code>-t</code>: Print wall-clock time taken by search phases.</li>
<li>
<code>-x &lt;bt2-idx&gt;</code>: Index file name prefix (minus
trailing .X.bt2).</li>
<li>
<code>-U</code>/<code>-1</code>,<code>-2</code>: Indicates the
location and names of the fastq files to align. <code>-U</code> stands
for single-end reads and <code>-1</code> and <code>-2</code> for the
pair 1 reads and pair 2 reads, respectively (paired-end reads).</li>
<li>
<code>-p</code>: Number of alignment threads to launch.</li>
<li>Additional parameters are added in the case of CUTandTAG samples,
following the analysis described in: <a href="https://yezhengstat.github.io/CUTTag_tutorial/#311_Alignment_to_HG38" class="external-link uri">https://yezhengstat.github.io/CUTTag_tutorial/#311_Alignment_to_HG38</a>
</li>
</ul>
<p>You can check the statistics of the alignment in the
<code>Logs/</code> folder and the result for the alignment can be found
in <code>BAM/*.raw.bam</code>.</p>
</div>
<div class="section level3">
<h3 id="post-processing">Post-processing<a class="anchor" aria-label="anchor" href="#post-processing"></a>
</h3>
<p>After the alignment, we need to do several post-processing steps in
order to have homogeneous and filtered data for downstream analyses.</p>
<ul>
<li>
<strong>Remove reads aligning to black-listed regions</strong>.</li>
<li>
<strong>Remove unaligned reads</strong>.</li>
<li>
<strong>Select reads aligning to cannonical
chromosomes</strong>.</li>
<li>
<strong>Remove duplicates</strong> using
<code>samtools markdup</code>.</li>
</ul>
<p>You can check the statistics in the <code>Logs/</code> folder and the
resulting file can be found in <code>BAM/*.bam</code>.</p>
</div>
<div class="section level3">
<h3 id="offset-correction-only-atac-seq">Offset correction (only ATAC-seq)<a class="anchor" aria-label="anchor" href="#offset-correction-only-atac-seq"></a>
</h3>
<p>If we are dealing with <strong>ATAC-seq</strong> data, we need to
perform an <strong>offset correction</strong> to correct the reads for
the transposase binding event. This is implemented in the function
<code><a href="../reference/offsetATAC.html">offsetATAC()</a></code>, which uses <a href="http://bedtools.readthedocs.io/en/latest/" class="external-link"><strong>bedtools</strong></a>
and <code>awk</code> and is used with paired-end data and
<code><a href="../reference/offsetATACSE.html">offsetATACSE()</a></code> which uses
<code><a href="https://rdrr.io/pkg/ATACseqQC/man/shiftGAlignments.html" class="external-link">ATACseqQC::shiftGAlignments()</a></code> to perform this same
correction (I don’t remember why I use different fuctions, I think the
ATACseqQC function did not work with paired-end reads?).</p>
<blockquote>
<p><strong>Warning</strong>: If you are trying to do some kind of
allelle-specific analysis or any other analysis in which it is important
the actual sequence of the read and its location, you might have to use
BAM files without the offset correction.</p>
</blockquote>
<p>The resulting file can be found in <code>BAM/*.offset.bam</code>.</p>
</div>
<div class="section level3">
<h3 id="peak-calling">Peak calling<a class="anchor" aria-label="anchor" href="#peak-calling"></a>
</h3>
<p>In order to proceed with the analysis it is key to have the
coordinates for the complete set of enriched regions in our experiment.
The process of searching and reporting such regions is called
<strong>peak calling</strong>.</p>
<p>To run these analysis we are going to use <a href="https://github.com/taoliu/MACS" class="external-link"><strong>MACS2</strong></a>
<code>callPeak</code> function, which is implemented in
<code><a href="../reference/callPeak.html">callPeak()</a></code>. The parameters used in this step are determined
by the argument <code>seq_type</code> but can be overwritten by passing
them to process_epigenome() using the arguments:</p>
<ul>
<li>
<code>type_peak</code> indicates whether to call <code>broad</code>
or <code>narrow</code> peaks.</li>
<li>
<code>shift</code> indicates whether to shift the called peaks using
the parameters <code>--shift -100 --extsize 200</code> in the MACS2
call.</li>
</ul>
<p>This function will return several files in the folder
<code>Peaks/</code> corresponding to our enriched regions.</p>
</div>
</div>
<div class="section level2">
<h2 id="generate-stats">Generate stats<a class="anchor" aria-label="anchor" href="#generate-stats"></a>
</h2>
<p>It is possible to summarize the stats from the processing using a
couple functions implemented in the package:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stats</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/getStats.html">getStats</a></span><span class="op">(</span>
    path_logs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Logs/"</span>, package<span class="op">=</span><span class="st">"pipelineNGS"</span><span class="op">)</span>,
    path_peaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Peaks/"</span>, package<span class="op">=</span><span class="st">"pipelineNGS"</span><span class="op">)</span>,
    peak_suffix <span class="op">=</span> <span class="st">"broadPeak"</span>
<span class="op">)</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="6%">
<col width="8%">
<col width="9%">
<col width="8%">
<col width="7%">
<col width="10%">
<col width="8%">
<col width="10%">
<col width="6%">
<col width="10%">
<col width="7%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left">sampleID</th>
<th align="right">total_reads</th>
<th align="right">aligned_reads</th>
<th align="right">multi_reads</th>
<th align="right">chrM_reads</th>
<th align="right">duplicate_reads</th>
<th align="right">final_reads</th>
<th align="right">alignment_rate</th>
<th align="right">chrM_rate</th>
<th align="right">duplicate_rate</th>
<th align="right">final_rate</th>
<th align="right">num_peaks</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HI19</td>
<td align="right">14558245</td>
<td align="right">14421420</td>
<td align="right">3069479</td>
<td align="right">81782</td>
<td align="right">816010</td>
<td align="right">11898585</td>
<td align="right">99.06015</td>
<td align="right">0.5617573</td>
<td align="right">5.60514</td>
<td align="right">81.73090</td>
<td align="right">50595</td>
</tr>
<tr class="even">
<td align="left">HI22</td>
<td align="right">48182652</td>
<td align="right">47910520</td>
<td align="right">10500399</td>
<td align="right">2859</td>
<td align="right">6792295</td>
<td align="right">36155652</td>
<td align="right">99.43521</td>
<td align="right">0.0059337</td>
<td align="right">14.09697</td>
<td align="right">75.03873</td>
<td align="right">80251</td>
</tr>
<tr class="odd">
<td align="left">HI37</td>
<td align="right">85131100</td>
<td align="right">84631826</td>
<td align="right">18313682</td>
<td align="right">5597</td>
<td align="right">18807614</td>
<td align="right">58077807</td>
<td align="right">99.41352</td>
<td align="right">0.0065746</td>
<td align="right">22.09253</td>
<td align="right">68.22161</td>
<td align="right">96752</td>
</tr>
<tr class="even">
<td align="left">HI40</td>
<td align="right">78309325</td>
<td align="right">78013102</td>
<td align="right">16958214</td>
<td align="right">4541</td>
<td align="right">13746501</td>
<td align="right">56648684</td>
<td align="right">99.62173</td>
<td align="right">0.0057988</td>
<td align="right">17.55410</td>
<td align="right">72.33964</td>
<td align="right">95067</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_linedraw</a></span><span class="op">(</span>base_size<span class="op">=</span><span class="fl">11</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Plot alignment rates</span>
<span class="va">al</span> <span class="op">&lt;-</span> 
  <span class="fu">pipelineNGS</span><span class="fu">::</span><span class="fu"><a href="../reference/plotAlignment.html">plotAlignment</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue3"</span>, <span class="st">"seagreen3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Alignment rate (%)"</span><span class="op">)</span> 

<span class="co">## Plot chrM rate</span>
<span class="va">chrm</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">stats</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sampleID</span>, <span class="va">chrM_rate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"chrM Rate (%)"</span><span class="op">)</span>

<span class="co">## Plot duplication rate</span>
<span class="va">dup</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">stats</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sampleID</span>, <span class="va">duplicate_rate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Duplication Rate (%)"</span><span class="op">)</span>

<span class="co">## Plot called peaks</span>
<span class="va">peaks</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">stats</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sampleID</span>, <span class="va">num_peaks</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">num_peaks</span><span class="op">)</span>, stat<span class="op">=</span><span class="st">"identity"</span>, color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="fl">0</span>, label<span class="op">=</span><span class="fu"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">(</span><span class="va">num_peaks</span>, accuracy<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>, hjust<span class="op">=</span><span class="fl">0</span>, nudge_y <span class="op">=</span> <span class="fl">0</span>, size<span class="op">=</span><span class="fl">3</span>,
             alpha<span class="op">=</span><span class="fl">.7</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels<span class="op">=</span><span class="va">comma</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"# of Peaks"</span><span class="op">)</span>

<span class="co">## Composite plot</span>
<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">al</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>, 
                   <span class="va">chrm</span>, 
                   <span class="va">dup</span>,
                   <span class="va">peaks</span>, 
                   ncol<span class="op">=</span><span class="fl">2</span>,
                   labels<span class="op">=</span><span class="st">"AUTO"</span><span class="op">)</span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="conversion-to-bigwig-and-visualization">Conversion to bigWig and Visualization<a class="anchor" aria-label="anchor" href="#conversion-to-bigwig-and-visualization"></a>
</h2>
<p>Once we have our post-processed data, we can proceed to visualize the
coverage of our samples in a genome browser. We can choose between
uploading the tracks to public servers and load them into <a href="https://genome.ucsc.edu/cgi-bin/hgTracks" class="external-link"><strong>UCSC Genome
Browser</strong></a> or to use a local browser like <a href="http://software.broadinstitute.org/software/igv/" class="external-link"><strong>IGV</strong></a>.</p>
<p>IGV allows to load BAM files and see the coverage and alignment of
the reads, but if we want to look just at the distribution of reads, we
need to create either BedGraph of BigWig files.</p>
<p><strong>WORK IN PROGRESS</strong></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
