<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='DEPRECATED! Use process_epigenome() with seq_type="CT" instead.'><title>Automated processing of CUT&amp;TAG samples — processCUTTAG • pipelineNGS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Automated processing of CUT&amp;TAG samples — processCUTTAG"><meta property="og:description" content='DEPRECATED! Use process_epigenome() with seq_type="CT" instead.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pipelineNGS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction.html">Processing epigenome data with pipelineNGS</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Automated processing of CUT&amp;TAG samples</h1>
      
      <div class="d-none name"><code>processCUTTAG.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>DEPRECATED! Use <code><a href="process_epigenome.html">process_epigenome()</a></code> with <code>seq_type="CT"</code> instead.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">processCUTTAG</span><span class="op">(</span>
  <span class="va">fastq_files</span>,
  <span class="va">out_name</span>,
  type <span class="op">=</span> <span class="st">"PE"</span>,
  run_fastqc <span class="op">=</span> <span class="cn">TRUE</span>,
  path_fastqc <span class="op">=</span> <span class="st">"FastQC/"</span>,
  path_bam <span class="op">=</span> <span class="st">"BAM/"</span>,
  path_peaks <span class="op">=</span> <span class="st">"Peaks/"</span>,
  path_logs <span class="op">=</span> <span class="st">"Logs/"</span>,
  index <span class="op">=</span> <span class="st">"/biodata/indices/species/Hsapiens/ucsc.hg19"</span>,
  remove <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrM"</span>, <span class="st">"chrUn"</span>, <span class="st">"_random"</span>, <span class="st">"_hap"</span>, <span class="st">"_gl"</span>, <span class="st">"EBV"</span><span class="op">)</span>,
  blacklist <span class="op">=</span> <span class="st">"~/data/consensusBlacklist.bed"</span>,
  gen_sizes <span class="op">=</span> <span class="st">""</span>,
  seacr_type <span class="op">=</span> <span class="st">"stingent"</span>,
  seacr_top <span class="op">=</span> <span class="fl">0.01</span>,
  cores <span class="op">=</span> <span class="fl">8</span>,
  bedtools_bamtobed <span class="op">=</span> <span class="st">"bedtools bamtobed"</span>,
  bedtools_genomecov <span class="op">=</span> <span class="st">"bedtools genomecov"</span>,
  seacr <span class="op">=</span> <span class="st">"SEACR_1.3.sh"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>fastq_files</dt>
<dd><p>List containing the different FastQ file pairs (one list
element for each sample).</p></dd>


<dt>out_name</dt>
<dd><p>Sample name to use for each of the samples.</p></dd>


<dt>type</dt>
<dd><p>Character indicating if reads are paired end (PE) or single end
(SE).</p></dd>


<dt>run_fastqc</dt>
<dd><p>Logical indicating whether to run FastQC or not.</p></dd>


<dt>path_fastqc</dt>
<dd><p>Path for the FastQC results.</p></dd>


<dt>path_bam</dt>
<dd><p>Path for the aligment and filtering results.</p></dd>


<dt>path_peaks</dt>
<dd><p>Path for the peak files.</p></dd>


<dt>path_logs</dt>
<dd><p>Path for the logs.</p></dd>


<dt>index</dt>
<dd><p>Path to the index needed by Bowtie2.</p></dd>


<dt>remove</dt>
<dd><p>List of chromosomes to remove from final BAM file.</p></dd>


<dt>blacklist</dt>
<dd><p>List of blaclisted regions to remove from BAM file.</p></dd>


<dt>gen_sizes</dt>
<dd><p>Chromosome sizes to use for bedgraph conversion.</p></dd>


<dt>seacr_type</dt>
<dd><p>Type of peaks to call with SEACR, either "stringent"
(default) or "relaxed."</p></dd>


<dt>seacr_top</dt>
<dd><p>A numeric threshold n between 0 and 1 returns the top n
fraction of peaks based on total signal within peaks. Default: 0.01.</p></dd>


<dt>cores</dt>
<dd><p>Number of threads to use for the analysis.</p></dd>


<dt>bedtools_bamtobed</dt>
<dd><p>Path or alias of the bedtools bamtobed utility.</p></dd>


<dt>bedtools_genomecov</dt>
<dd><p>Path or alias of the bedtools genomecov utility.</p></dd>


<dt>seacr</dt>
<dd><p>Path or alias of the SEACR utilty.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

