#' Get alignment and processing stats from BAM files
#'
#' Given a single or multiple BAM files outputed by alignmentBowtie2() and filtOutBAM()
#' returns a data.frame with a summary of alignment and post-processing statistics.
#' @param raw_bam Character vector or string of BAM files generated by alignmentBowtie2() with the suffix ".raw".
#' @param path_logs Path where the log files from samtools mrkdup are stored.
#' @param path_peks Path where the peaks are saved.
#' @param peak_type Type of the peak files, either narrowPeak or broadPeak
#' @return Data.frame with some alignment and processing statistics for all the input samples.
#' @export
#' @examples
#' \dontrun{
#' stats <- getStats(raw_bam=c("path/to/file.bam", "path/to/file2.bam"), path_logs="path/logs")
#' }
getStats <- function(path_logs,
                     path_peaks,
                     peak_type="narrowPeak") {

  names <- gsub(".alignment.log", "", list.files(path_logs, pattern="alignment.log"))
  stats <- lapply(names,
                  .getStatsSingle,
                  path_logs=path_logs,
                  path_peaks=path_peaks,
                  peak_type=peak_type)
  stats <- as.data.frame(do.call(rbind, stats),
                         stringsAsFactors=FALSE)
  return(stats)
}



.getStatsSingle <- function(name,
                            path_logs,
                            path_peaks,
                            peak_type) {

  # Alignment stats from log file
  stats <- .parseBowtie2LogSE(file.path(path_logs, paste0(name, ".alignment.log")))
  total_reads <- stats$unpaired
  total_aligned <- stats$unique + stats$multi
  alignment_rate <- total_aligned/total_reads*100

  # Chr stats
  idxstats <- tableFromIdxstats(file.path(path_logs, paste0(name, ".raw.idxstats.log")))
  chrM <- idxstats$aligned[idxstats$chr=="chrM"]

  # Get duplicates stats
  if (file.exists(file.path(path_logs, paste0(name, ".rmdup.log")))) {
    dup.tab <- read.delim(file.path(path_logs, paste0(name, ".rmdup.log")),
                          stringsAsFactors=FALSE, header=FALSE)
    duplicates <- as.numeric(strsplit(dup.tab[grep("DUPLICATE TOTAL", dup.tab[,1]),1], ": ")[[1]][2])
  } else {
    duplicates <- NA
  }

  # Get number of peaks called
  files <- list.files(path_peaks,
                      pattern=peak_type,
                      full.names=T)
  files <- files[grep(paste0(name, "_"), files)]

  if (length(files)==1) {
    peaks <- as.numeric(strsplit(system(paste("wc -l", files), intern = T), " ")[[1]][1])
  } else {
    peaks <- NA
  }

  # Get final number of reads
  stats_final <-tableFromIdxstats(file.path(path_logs, paste0(name, ".idxstats.log")))
  final_reads <- sum(stats_final$aligned)

  # Create data.frame
  df <- data.frame(sampleID=name,
                   total_reads=total_reads,
                   aligned_reads=total_aligned,
                   multi_reads=stats$multi,
                   chrM_reads=chrM,
                   duplicate_reads=duplicates,
                   final_reads=final_reads,
                   alignment_rate=total_aligned/total_reads*100,
                   chrM_rate=chrM/total_reads*100,
                   duplicate_rate=duplicates/total_reads*100,
                   final_rate=final_reads/total_reads*100,
                   num_peaks=peaks)
}

.parseBowtie2LogSE <- function(logfile, paired=FALSE) {
  log <- readLines(con = logfile)
  unpaired <- .obtainReadNumber(log[grep("were unpaired", log)])
  unaligned <- .obtainReadNumber(log[grep("aligned 0 times", log)])
  unique <- .obtainReadNumber(log[grep("aligned exactly 1 time", log)])
  multi <- .obtainReadNumber(log[grep("aligned >1 times", log)])

  stats <- data.frame(unpaired, unaligned, unique, multi)
  return(stats)
}

.obtainReadNumber <- function(line) {
  as.numeric(strsplit(gsub(" ", "", line), "(", fixed=TRUE)[[1]][1])
}

#' Table form idxstats
#'
#' Given a bam file, returns samtools idxstats output as a data.frame.
#' @param bam Character string with the name and path of a single BAM file. Index should
#' be in the same folder.
#' @return A data.frame with the output of samtools idxstats.
#' @export
#' @examples
#' \dontrun{
#' df <- tableFromIdxstats("path/to/file.bam")
#' }
tableFromIdxstats <- function(log) {
  stats <- read.delim(log)
  colnames(stats) <- c("chr", "chr_len", "aligned", "not_aligned")
  stats$chr_len <- as.numeric(stats$chr_len)
  stats$aligned <- as.numeric(stats$aligned)
  stats$not_aligned <- as.numeric(stats$not_aligned)
  return(stats)
}
