#' Get alignment and processing stats from BAM files
#'
#' Given a single or multiple BAM files outputed by alignmentBowtie2() and filtOutBAM()
#' returns a data.frame with a summary of alignment and post-processing statistics.
#' @param raw_bam Character vector or string of BAM files generated by alignmentBowtie2() with the suffix ".raw".
#' @param path_logs Path where the log files from samtools mrkdup are stored.
#' @return Data.frame with some alignment and processing statistics for all the input samples.
#' @export
#' @examples
#' \dontrun{
#' stats <- getStats(raw_bam=c("path/to/file.bam", "path/to/file2.bam"), path_logs="path/logs")
#' }
getStats <- function(raw_bam,
                     path_logs) {
  stats <- lapply(raw_bam,
                  .getStatsSingle,
                  path_logs=path_logs)
  stats <- as.data.frame(do.call(rbind, stats),
                         stringsAsFactors=FALSE)
  return(stats)
}

#' Table form idxstats
#'
#' Given a bam file, returns samtools idxstats output as a data.frame.
#' @param bam Character string with the name and path of a single BAM file. Index should
#' be in the same folder.
#' @return A data.frame with the output of samtools idxstats.
#' @export
#' @examples
#' \dontrun{
#' df <- tableFromIdxstats("path/to/file.bam")
#' }
tableFromIdxstats <- function(bam) {
  stats <- system(paste("samtools idxstats", bam),
                  intern = TRUE)

  stats <- as.data.frame(do.call(rbind, strsplit(stats, "\t")),
                         stringsAsFactors=FALSE)
  colnames(stats) <- c("chr", "chr_len", "aligned", "not_aligned")
  stats$chr_len <- as.numeric(stats$chr_len)
  stats$aligned <- as.numeric(stats$aligned)
  stats$not_aligned <- as.numeric(stats$not_aligned)
  return(stats)
}

.getStatsSingle <- function(bam,
                            path_logs) {
  # Stats for RAW file
  stats <- tableFromIdxstats(bam)

  name <- getNameFromPath(bam, suffix=".raw.bam")
  total_reads <- sum(stats$aligned) + sum(stats$not_aligned)
  total_aligned <- sum(stats$aligned)
  alignment_rate <- total_aligned/total_reads*100
  chrM <- stats$aligned[stats$chr=="chrM"]

  # Get duplicates stats
  if (file.exists(paste0(path_logs, name, ".rmdup.log"))) {
    dup.tab <- read.delim(paste0(path_logs, name, ".rmdup.log"),
                          stringsAsFactors=FALSE, header=FALSE)[5,]
    duplicates <- as.numeric(strsplit(dup.tab, " ")[[1]][3])
  } else {
    duplicates <- NA
  }

  # Get final number of reads
  stats_final <- tableFromIdxstats(gsub(".raw", "", bam))
  final_reads <- sum(stats_final$aligned)

  # Create data.frame
  df <- data.frame(sampleID=name,
                   total_reads=total_reads,
                   aligned_reads=total_aligned,
                   chrM_reads=chrM,
                   duplicate_reads=duplicates,
                   final_reads=final_reads,
                   alignment_rate=total_aligned/total_reads*100,
                   chrM_rate=chrM/total_reads*100,
                   duplicate_rate=duplicates/total_reads*100,
                   final_rate=final_reads/total_reads*100)
}
